loopback.go

```diff
@@ -44,20 +44,75 @@ type FS struct {
 
 	xlock  sync.RWMutex
 	xattrs map[string]map[string][]byte
+
+	nlock sync.Mutex
+	nodes map[string][]*Node // realPath -> nodes
 }
 
 func newFS(rootPath string) *FS {
 	return &FS{
 		rootPath: rootPath,
 		xattrs:   make(map[string]map[string][]byte),
+		nodes:    make(map[string][]*Node),
+	}
+}
+
+func (f *FS) newNode(n *Node) {
+	rp := n.getRealPath()
+
+	f.nlock.Lock()
+	defer f.nlock.Unlock()
+	if f.nodes[rp] == nil {
+		f.nodes[rp] = make([]*Node, 0)
+	}
+	f.nodes[rp] = append(f.nodes[rp], n)
+}
+
+func (f *FS) nodeRenamed(oldPath string, newPath string) {
+	f.nlock.Lock()
+	defer f.nlock.Unlock()
+	f.nodes[newPath] = append(f.nodes[newPath], f.nodes[oldPath]...)
+	for _, n := range f.nodes[newPath] {
+		n.updateRealPath(newPath)
+	}
+}
+
+func (f *FS) forgetNode(n *Node) {
+	f.nlock.Lock()
+	defer f.nlock.Unlock()
+	nodes, ok := f.nodes[n.realPath]
+	if !ok {
+		return
+	}
+
+	found := -1
+	for i, node := range nodes {
+		if node == n {
+			found = i
+			break
+		}
+	}
+
+	if found > -1 {
+		for i := found; i < len(nodes)-1; i++ {
```

Comment:
Whoa ... I thought `[]int{1}[1:]` would cause an "index out of range" error but just tried it and it worked! And it's in the specification:

> For arrays or strings, the indices are in range if 0 <= low <= high <= len(a), otherwise they are out of range.

Thanks!